pkg/codegen/convert/plugin_mapper.go

```diff
@@ -173,12 +173,96 @@ func NewPluginMapper(ws Workspace,
+		// things like providers failing to start.
+		return nil, "", fmt.Errorf("could not create provider '%s': %w", pluginSpec.name, err)
+	}
+	defer contract.IgnoreClose(providerPlugin)
+
+	data, mappedProvider, err := providerPlugin.GetMapping(l.conversionKey)
+	if err != nil {
+		// This was an error calling GetMapping, not just that GetMapping returned a nil result. It's fine for
+		// GetMapping to return (nil, nil) as that simply indicates that the plugin doesn't have a mapping for
+		// the requested key.
+		return nil, "", fmt.Errorf("could not get mapping for provider '%s': %w", pluginSpec.name, err)
+	}
+	// A provider returns empty if it didn't have a mapping
+	if mappedProvider != "" && len(data) != 0 {
+		// If this was the provider we we're looking for we can now return it
+		if mappedProvider == targetProvider {
+			return data, mappedProvider, nil
+		}
+	}
+	// TODO: Temporary hack to work around the fact that most of the plugins return a mapping for "tf" but
+	// not "terraform" but they're the same thing.
+	if l.conversionKey == "terraform" {
+		// Copy-pasta of the above _but_ we'll delete this whole if block once the plugins have had a
+		// chance to update.
+		data, mappedProvider, err = providerPlugin.GetMapping("tf")
+		if err != nil {
+			// This was an error calling GetMapping, not just that GetMapping returned a nil result. It's fine
+			// for GetMapping to return (nil, nil) as that simply indicates that the plugin doesn't have a
+			// mapping for the requested key.
+			return nil, "", fmt.Errorf("could not get mapping for provider '%s': %w", pluginSpec.name, err)
+		}
+		// A provider returns empty if it didn't have a mapping
+		if mappedProvider != "" && len(data) != 0 {
+			// If this was the provider we we're looking for we can now return it
+			if mappedProvider == targetProvider {
+				return data, mappedProvider, nil
+			}
+		}
+	}
+
+	return data, mappedProvider, err
+}
+
 func (l *pluginMapper) GetMapping(provider string) ([]byte, error) {
-	// Do we already have an entry for this provider, if so use it
+	// If we already have an entry for this provider, use it
 	if entry, has := l.entries[provider]; has {
 		return entry, nil
 	}
 
+	// No entry yet, we're going to _try_ to get the plugin that matches the provider name as generally these
+	// will match up and it saves us doing a linear search through all plugins.
+	matchIdx := -1
+	for i, pluginSpec := range l.plugins {
+		if pluginSpec.name == tokens.Package(provider) {
+			matchIdx = i
+			break
+		}
+	}
+
+	if matchIdx != -1 {
+		// Pop this out the list and try and get its mapping. We pop by doing a swap and remove of the last
+		// element as order doesn't matter here.
+		pluginSpec := l.plugins[matchIdx]
+		l.plugins[matchIdx] = l.plugins[len(l.plugins)-1]
+		l.plugins = l.plugins[:len(l.plugins)-1]
+
+		data, mappedProvider, err := l.getMappingForPlugin(provider, pluginSpec)
+		if err != nil {
+			return nil, err
```

Comment:
also, taking a second look, we don't need to duplicate this at all.
we can just move `l.plugins[matchIdx]` to the front of the list and get the same optimization.

wil submit a fix.

Reply:
pushed fix